crates/assembly/src/basic_block_builder.rs

```diff
@@ -161,6 +165,14 @@ impl BasicBlockBuilder<'_> {
         } else {
             assembly_op.set_num_cycles(cycle_count as u8);
 
+            // Create a 1-1 mapping: link the AsmOp to all operations it covers.
+            // The first operation is already linked at op_start, so we add links for
+            // operations from op_start + 1 to op_start + cycle_count - 1.
+            // We insert them in reverse order to maintain the sorted order of the decorators list.
+            for op_idx in ((op_start + 1)..(op_start + cycle_count)).rev() {
+                self.decorators.insert(self.last_asmop_pos + 1, (op_idx, assembly_op_id));
```

Comment:
This line reuses the same `assembly_op_id` for all operations covered by the AsmOp. That's the correct approach per issue #2446 - we're linking multiple operations to one decorator, not duplicating the decorator itself.

For a 3-cycle instruction starting at operation 5, this creates:
```
decorators = [..., (5, id), (6, id), (7, id), ...]
```
All three entries share the same decorator ID, so no memory duplication of the AsmOp itself.

The reverse insertion order maintains the sorted decorator list without extra sorting cost.